qiskit_experiments/library/randomized_benchmarking/sampling_utils.py

```diff
@@ -0,0 +1,359 @@
+
+        layers = []
+        for samplelayer in samples:
+            layers.append(
+                tuple(GateInstruction(*ins) for ins in zip(((j,) for j in qubits), samplelayer))
+            )
+        return layers
+
+
+class EdgeGrabSampler(BaseSampler):
+    r"""A sampler that uses the edge grab algorithm [1] for sampling gate layers.
+
+    The edge grab sampler, given a list of :math:`w` qubits, their connectivity
+    graph, and the desired two-qubit gate density :math:`\xi_s`, outputs a layer
+    as follows:
+
+    1. Begin with the empty set :math:`E` and :math:`E_r`, the set of all edges
+    in the connectivity graph. Select an edge from :math:`E_r` at random and
+    add it to :math:`E`, removing all edges that share a qubit with the edge
+    from :math:`E_r`.
+
+    2. Select edges from :math:`E` with the probability :math:`w\xi/2|E|`. These
+    edges will have two-qubit gates in the output layer.
+
+    |
+
+    This produces a layer with an expected two-qubit gate density :math:`\xi`. In the
+    default mirror RB configuration where these layers are dressed with single-qubit
+    Pauli layers, this means the overall expected two-qubit gate density will be
+    :math:`\xi_s/2=\xi`. The actual density will converge to :math:`\xi_s` as the
+    circuit size increases.
+
+    .. ref_arxiv:: 1 2008.11294
+
+    """
+
+    def __call__(
+        self,
+        qubits: Sequence,
+        length: int = 1,
+    ) -> List[Tuple[GateInstruction]]:
+        """Sample layers using the edge grab algorithm.
+
+        Args:
+            qubits: A sequence of qubits to generate layers for.
+            length: The length of the sequence to output.
+            seed: Seed for random generation.
+
+        Raises:
+            Warning: If the coupling map has no connectivity or
+                ``two_qubit_gate_density`` is too high.
+            TypeError: If invalid gate set(s) are specified.
+
+        Returns:
+            A ``length``-long list of :class:`qiskit.circuit.QuantumCircuit` layers over
+            ``num_qubits`` qubits. Each layer is represented by a list of tuples which
+            are in the format ((one or more qubit indices), gate). Single-qubit
+            Cliffords are represented by integers for speed. Here's an example with the
+            default choice of Cliffords for the single-qubit gates and CXs for the
+            two-qubit gates:
+
+            .. parsed-literal::
+                (((1, 2), CXGate()), ((0,), 12), ((3,), 20))
+
+            This represents a layer where the 12th Clifford is performed on qubit 0,
+            a CX is performed with control qubit 1 and target qubit 2, and the 20th
+            Clifford is performed on qubit 3.
+
+        """
+        num_qubits = len(qubits)
+        gateset = self._probs_by_gate_size()
+        norm1q = sum(gateset[1][1])
+        norm2q = sum(gateset[2][1])
+        two_qubit_gate_density = sum(gateset[2][1]) / (sum(gateset[2][1]) + sum(gateset[1][1]))
```

Comment:
This happens if gates that span more than 2 qubits are included in the distribution, but in this case I should raise an error instead of continuing. Updated.